### ISSUE TITLE
6.3.0 breaks deserialization of DateRange fields

### ISSUE THREAD

**Issue by seanthegeek** [elastic/elasticsearch-dsl-py#1074]
```
File "C:\Users\Sean\GitHub\parsedmarc\parsedmarc\elastic.py", line 94, in save
    return super().save(** kwargs)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\document.py", line 425, in save
    self.full_clean()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 491, in full_clean
    self.clean_fields()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 477, in clean_fields
    data = field.clean(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 99, in clean
    data = self.deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 95, in deserialize
    return self._deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 387, in _deserialize
    return Range(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\wrappers.py", line 21, in __init__
    raise ValueError('Range accepts a single dictionary or a set of keyword arguments.')
ValueError: Range accepts a single dictionary or a set of keyword arguments.
```

**Comment by HonzaKral** [comment]
Thank you for the report, can you please share what is the value you are using for the range field? The raw json would be ideal.

Thanks!

**Comment by seanthegeek** [comment]
Here's the raw JSON from Elasticsearch:

```json
{
  "_index": "dmarc_aggregate-2018-10-01",
  "_type": "doc",
  "_id": "3hEkV2cBXjh0np_JFRUh",
  "_version": 1,
  "_score": 2,
  "_source": {
    "xml_schemea": "draft",
    "org_name": "accurateplastics.com",
    "org_email": "administrator@accurateplastics.com",
    "report_id": "example.com:1538463741",
    "date_range": [
      "2018-10-01T13:07:12",
      "2018-10-01T13:07:12"
    ],
    "published_policy": {
      "domain": "example.com",
      "adkim": "r",
      "aspf": "r",
      "p": "none",
      "sp": "reject",
      "pct": 100,
      "fo": "0"
    },
    "source_ip_address": "12.20.127.122",
    "source_country": "US",
    "source_reverse_dns": "smtp.cardhealth.com",
    "source_base_domain": "cardhealth.com",
    "message_count": 1,
    "disposition": "none",
    "dkim_aligned": false,
    "spf_aligned": false,
    "header_from": "example.com",
    "spf_results": [
      {
        "scope": "mfrom",
        "result": "none"
      }
    ],
    "passed_dmarc": false
  },
  "fields": {
    "date_range": [
      "2018-10-01T13:07:12.000Z",
      "2018-10-01T13:07:12.000Z"
    ]
  }
}
```

And here are the Python class definitions:

```python
from elasticsearch_dsl import connections, Object, Document, Index, Nested, \
    InnerDoc, Integer, Text, Boolean, DateRange, Ip, Date, Search

class _PolicyOverride(InnerDoc):
    type = Text()
    comment = Text()


class _PublishedPolicy(InnerDoc):
    domain = Text()
    adkim = Text()
    aspf = Text()
    p = Text()
    sp = Text()
    pct = Integer()
    fo = Text()


class _DKIMResult(InnerDoc):
    domain = Text()
    selector = Text()
    result = Text()


class _SPFResult(InnerDoc):
    domain = Text()
    scope = Text()
    results = Text()


class _AggregateReportDoc(Document):
    class Index:
        name = "dmarc_aggregate"

    xml_schema = Text()
    org_name = Text()
    org_email = Text()
    org_extra_contact_info = Text()
    report_id = Text()
    date_range = DateRange()
    errors = Text()
    published_policy = Object(_PublishedPolicy)
    source_ip_address = Ip()
    source_country = Text()
    source_reverse_dns = Text()
    source_Base_domain = Text()
    message_count = Integer
    disposition = Text()
    dkim_aligned = Boolean()
    spf_aligned = Boolean()
    passed_dmarc = Boolean()
    policy_overrides = Nested(_PolicyOverride)
    header_from = Text()
    envelope_from = Text()
    envelope_to = Text()
    dkim_results = Nested(_DKIMResult)
    spf_results = Nested(_SPFResult)

    def add_policy_override(self, type_, comment):
        self.policy_overrides.append(_PolicyOverride(type=type_,
                                                     comment=comment))

    def add_dkim_result(self, domain, selector, result):
        self.dkim_results.append(_DKIMResult(domain=domain,
                                             selector=selector,
                                             result=result))

    def add_spf_result(self, domain, scope, result):
        self.spf_results.append(_SPFResult(domain=domain,
                                           scope=scope,
                                           result=result))

    def save(self, ** kwargs):
        self.passed_dmarc = False
        self.passed_dmarc = self.spf_aligned or self.dkim_aligned

        return super().save(** kwargs)


class _EmailAddressDoc(InnerDoc):
    display_name = Text()
    address = Text()


class _EmailAttachmentDoc(Document):
    filename = Text()
    content_type = Text()


class _ForensicSampleDoc(InnerDoc):
    raw = Text()
    headers = Object()
    headers_only = Boolean()
    to = Nested(_EmailAddressDoc)
    subject = Text()
    filename_safe_subject = Text()
    _from = Object(_EmailAddressDoc)
    date = Date()
    reply_to = Nested(_EmailAddressDoc)
    cc = Nested(_EmailAddressDoc)
    bcc = Nested(_EmailAddressDoc)
    body = Text()
    attachments = Nested(_EmailAttachmentDoc)

    def add_to(self, display_name, address):
        self.to.append(_EmailAddressDoc(display_name=display_name,
                                        address=address))

    def add_reply_to(self, display_name, address):
        self.reply_to.append(_EmailAddressDoc(display_name=display_name,
                                              address=address))

    def add_cc(self, display_name, address):
        self.cc.append(_EmailAddressDoc(display_name=display_name,
                                        address=address))

    def add_bcc(self, display_name, address):
        self.bcc.append(_EmailAddressDoc(display_name=display_name,
                                         address=address))

    def add_attachment(self, filename, content_type):
        self.attachments.append(_EmailAttachmentDoc(filename=filename,
                                content_type=content_type))
```

**Comment by HonzaKral** [comment]
Thank you, the value for `date_range` that your document has (`["2018-10-01T13:07:12", "2018-10-01T13:07:12"]`) is not a valid value for a `date_range` type field, when you try to index it you get:

```json
{
  "error": {
    "root_cause": [
      {
        "type": "mapper_parsing_exception",
        "reason": "error parsing field [date_range], expected an object but got null"
      }
    ],
    "type": "mapper_parsing_exception",
    "reason": "failed to parse field [date_range] of type [date_range]",
    "caused_by": {
      "type": "mapper_parsing_exception",
      "reason": "error parsing field [date_range], expected an object but got null"
    }
  },
  "status": 400
}
```

Can you please verify that the mappings in elasticsearch is indeed set to `date_range` for that field and that the document is properly indexed? Thank you!

**Comment by seanthegeek** [comment]
That was a bad example. Here's a better one:

```json
{
  "_index": "dmarc_aggregate-2018-10-05",
  "_type": "doc",
  "_id": "2xEkV2cBXjh0np_JEhVl",
  "_version": 1,
  "_score": 2,
  "_source": {
    "xml_schemea": "1.0",
    "org_name": "usssa.com",
    "org_email": "postmaster@usssa.com",
    "report_id": "8953b4d4a4ee4218b6ac0e2cb2667ee1",
    "date_range": [
      "2018-10-05T20:00:00",
      "2018-10-06T19:59:59"
    ],
    "published_policy": {
      "domain": "example.com",
      "adkim": "r",
      "aspf": "r",
      "p": "none",
      "sp": "none",
      "pct": 100,
      "fo": "0"
    },
    "source_ip_address": "12.20.127.40",
    "source_country": "US",
    "source_reverse_dns": "smtp3.cardinal.com",
    "source_base_domain": "cardinal.com",
    "message_count": 1,
    "disposition": "none",
    "dkim_aligned": false,
    "spf_aligned": false,
    "header_from": "example.com",
    "passed_dmarc": false
  },
  "fields": {
    "date_range": [
      "2018-10-05T20:00:00.000Z",
      "2018-10-06T19:59:59.000Z"
    ]
  }
}```

The mapping type is `date`, not `date_range`, even though `DateRange()` is used in the class definition, but this has worked fine until elasticsearch-dsl 6.3.0.

Mapping:

```json
{
  "mapping": {
    "doc": {
      "properties": {
        "date_range": {
          "type": "date"
        },
        "disposition": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "dkim_aligned": {
          "type": "boolean"
        },
        "header_from": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "message_count": {
          "type": "long"
        },
        "org_email": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "org_name": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "passed_dmarc": {
          "type": "boolean"
        },
        "published_policy": {
          "properties": {
            "adkim": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "aspf": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "domain": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "fo": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "p": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            },
            "pct": {
              "type": "long"
            },
            "sp": {
              "type": "text",
              "fields": {
                "keyword": {
                  "type": "keyword",
                  "ignore_above": 256
                }
              }
            }
          }
        },
        "report_id": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "source_base_domain": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "source_country": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "source_ip_address": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "source_reverse_dns": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        },
        "spf_aligned": {
          "type": "boolean"
        },
        "xml_schemea": {
          "type": "text",
          "fields": {
            "keyword": {
              "type": "keyword",
              "ignore_above": 256
            }
          }
        }
      }
    }
  }
}
```

**Comment by HonzaKral** [comment]
Well, the problem is that your `python` code doesn't match the reality of what's your index, it is not a range field. The change in `6.3` just exposed it.

The fix is to recreate your index with the proper mappings (by calling the `.init()` classmethod). The workaround, to get it working, is to change your python code to just have `Date` and not `DateRange`.

**Comment by seanthegeek** [comment]
Thanks. I'll use the workaround. I'm a little confused about how the mismatch happened.

The indexes and mappings were created by those Python classes, so is elasticsearch-dsl using the wrong mapping type for DateRange field objects? If not, how could a mismatch happen?

**Comment by seanthegeek** [comment]
Changing the Python code from `DateRange` to `Date` results in another exception

> not all arguments converted during string formatting

**Comment by seanthegeek** [comment]
I se what's happening. `Date` expects a single `datetime` object, where I'm providing a tuple:

```python
 (datetime.datetime(2018, 1, 15, 19, 0), datetime.datetime(2018, 1, 16, 18, 59, 59))
```

Full traceback

```python
  File "C:\Users\Sean\GitHub\parsedmarc\parsedmarc\elastic.py", line 94, in save
    return super().save(** kwargs)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\document.py", line 425, in save
    self.full_clean()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 491, in full_clean
    self.clean_fields()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 477, in clean_fields
    data = field.clean(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 99, in clean
    data = self.deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 95, in deserialize
    return self._deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 259, in _deserialize
    raise ValidationException('Could not parse date from the value (%r)' % data)
TypeError: not all arguments converted during string formatting
```
So then I tried 6.3.0 with the example from https://elasticsearch-dsl.readthedocs.io/en/6.3.0/persistence.html#data-types, and I ran into the same exception that led me to open this issue

```python
from datetime import  datetime

from elasticsearch_dsl import Document, DateRange, Keyword, Range

class RoomBooking(Document):
    class Index:
        name = "test"
    room = Keyword()
    dates = DateRange()


rb = RoomBooking(
  room='Conference Room II',
  dates=Range(
    gte=datetime(2018, 11, 17, 9, 0, 0),
    lt=datetime(2018, 11, 17, 10, 0, 0)
  )
)

rb.save()
```

```python
Traceback (most recent call last):
  File "C:/Users/Sean/.PyCharm2018.3/config/scratches/test_dsl.py", line 20, in <module>
    rb.save()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\document.py", line 425, in save
    self.full_clean()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 491, in full_clean
    self.clean_fields()
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\utils.py", line 477, in clean_fields
    data = field.clean(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 99, in clean
    data = self.deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 95, in deserialize
    return self._deserialize(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\field.py", line 387, in _deserialize
    return Range(data)
  File "C:\Users\Sean\GitHub\parsedmarc\venv\lib\site-packages\elasticsearch_dsl\wrappers.py", line 21, in __init__
    raise ValueError('Range accepts a single dictionary or a set of keyword arguments.')
ValueError: Range accepts a single dictionary or a set of keyword arguments.
```

**Comment by HonzaKral** [comment]
Thank you for your patience @seanthegeek , could you please verify that 600eff6 fixed the issue for you so we can release `6.3.1`?

**Comment by seanthegeek** [comment]
Confirmed fixed. Now I just need to figure out how to reindex all of the indexes (I split them by date)  so tha they use date_range instead of date.

**Comment by seanthegeek** [comment]
On second thought...can you refactor Date so that it also accepts a tuple of datetime, like the old release did? WhenRange is used, Kibana is wanting to break up the date range into to separate fields, which is not what is wanted.

![image](https://user-images.githubusercontent.com/44679/49191909-a1a4ea00-f346-11e8-9066-7676cd0b839d.png)

**Comment by HonzaKral** [comment]
`tuple`s were never really supported but should be now with 2880320474d1f5ddaa1ba44542210970bf19c065, if there are no issues I will cut a `6.3.1` release soon

**Comment by seanthegeek** [comment]
That didn't fix it, but I fixed it in PR #1078 :)
